<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<head>
<!-- 2017-11-02 Thu 15:39 -->
<meta http-equiv="Content-Type" content="text/html;charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<title>Fine mapping idea</title>
<meta name="generator" content="Org mode" />
<meta name="author" content="Abhishek Sarkar" />
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link rel="stylesheet" type="text/css" href="https://www.pirilampo.org/styles/readtheorg/css/htmlize.css"/>
<link rel="stylesheet" type="text/css" href="https://www.pirilampo.org/styles/readtheorg/css/readtheorg.css"/>
<script src="https://ajax.googleapis.com/ajax/libs/jquery/2.1.3/jquery.min.js"></script>
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.4/js/bootstrap.min.js"></script>
<script type="text/javascript" src="http://www.pirilampo.org/styles/lib/js/jquery.stickytableheaders.min.js"></script>
<script type="text/javascript" src="http://www.pirilampo.org/styles/readtheorg/js/readtheorg.js"></script>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Fine mapping idea</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org76749e6">Proposed model</a></li>
<li><a href="#orgd8882f1">Variational lower bound</a></li>
<li><a href="#orgeee1ab8">Setup</a></li>
<li><a href="#org6776455">Simulated data</a></li>
<li><a href="#orgae34bdc">Coordinate ascent</a></li>
<li><a href="#orge47c8d0">Harder simulation</a></li>
</ul>
</div>
</div>

<div id="outline-container-org76749e6" class="outline-2">
<h2 id="org76749e6">Proposed model</h2>
<div class="outline-text-2" id="text-org76749e6">
<p>
Revise the notation, using Latin for the model and Greek for the variational
approximation.
</p>

<p>
\[ p(\mathbf{y} \mid \mathbf{x}, \mathbf{w}) = N(\mathbf{y}; \mathbf{x} \mathbf{w}', v^{-1} \mathbf{I}) \]
\[ \mathbf{w} = \sum_k z_{kj} b_{kj} \]
</p>

<p>
\[ p(b_{kj} \mid z_k = 1, v, v_b) = N(b_{kj}; 0, v^{-1} v_b^{-1}) \]
\[ p(b_{kj} \mid z_k = 0) = \delta(b_{kj}) \]
</p>

<p>
\[ p(z_k \mid \mathbf{p}) = \mathrm{Multinomial}(z_k; 1, \mathbf{p}) \]
</p>

<p>
\[ q(b_{kj} \mid z_{kj} = 1, \mu, \phi) = N(b_{kj}; \mu_{kj}, \phi_{kj}^{-1}) \]
\[ q(b_{kj} \mid z_{kj} = 0, \mu, \phi) = \delta(b_{kj}) \]
</p>

<p>
\[ q(z_k \mid \pi) = \mathrm{Multinomial}(z_k; 1, \mathbf{\pi}_k) \]
</p>
</div>
</div>

<div id="outline-container-orgd8882f1" class="outline-2">
<h2 id="orgd8882f1">Variational lower bound</h2>
<div class="outline-text-2" id="text-orgd8882f1">
<p>
\[ \mathrm{ELBO} = \mathbb{E}_q[\ln p(\mathbf{y} \mid \mathbf{x}, \mathbf{w})] -
  KL(q(\mathbf{z}) \Vert p(\mathbf{z})) - KL(q(\mathbf{b} \mid \mathbf{z}) \Vert p(\mathbf{b} \mid \mathbf{z}))
  \]
\[ \mathbb{E}_q[\ln p(\mathbf{y} \mid \mathbf{x}, \mathbf{w})] = -\frac{v}{2} \sum_i
  \left(y_i - \sum_j x_{ij} \mathbb{E}_q[w_j]\right)^2 - \sum_{i,j} x_{ij}^2 \mathbb{V}_q[w_j] \]
</p>

<p>
\[ \mathbb{E}_q[w_j] = \sum_k \pi_{kj} \mu_{kj} \]
\[ \mathbb{V}_q[w_j] = \sum_k \pi_{kj} \phi_{kj}^{-1} + \pi_{kj} (1 - \pi_{kj}) \mu_{kj}^2 \]
</p>

<p>
\[ KL(q(\mathbf{z}) \Vert p(\mathbf{z})) = \sum_k E_q[\ln q(z_k) - \ln p(z_k)] \]
\[ = \sum_k \left[\sum_j \pi_{kj} \ln \pi_{kj} - \sum_j \pi_{kj} \ln p_{j}\right] \]
\[ = \sum_{j,k} \pi_{kj} \left( \ln \pi_{kj} - \ln p_{j} \right) \]
</p>

<p>
\[ KL(q(\mathbf{b} \mid \mathbf{z}) \Vert p(\mathbf{b} \mid \mathbf{z})) =
  \frac{1}{2} \sum_{j,k} \pi_{kj} \left(1 + \ln (v v_b) - \ln \phi_{kj} + v v_b (\mu_{kj}^2 + \phi_{kj}^{-1}) \right)\]
</p>
</div>
</div>

<div id="outline-container-orgeee1ab8" class="outline-2">
<h2 id="orgeee1ab8">Setup</h2>
<div class="outline-text-2" id="text-orgeee1ab8">
<div class="org-src-container">
<pre class="src src-shell" id="orgd017887">sbatch --partition=broadwl --mem=16G --time=36:00:00 --job-name=ipython3 --output=ipython3.out
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate nwas
rm -f $<span class="org-variable-name">HOME</span>/.local/share/jupyter/runtime/kernel-aksarkar.json
ipython3 kernel --ip=$(<span class="org-sh-quoted-exec">hostname</span> -i) -f kernel-aksarkar.json
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython" id="org389d913">%matplotlib inline
<span class="org-keyword">import</span> edward <span class="org-keyword">as</span> ed
<span class="org-keyword">import</span> matplotlib.pyplot <span class="org-keyword">as</span> plt
<span class="org-keyword">import</span> nwas
<span class="org-keyword">import</span> numpy <span class="org-keyword">as</span> np
<span class="org-keyword">import</span> pandas <span class="org-keyword">as</span> pd
<span class="org-keyword">import</span> pyplink
<span class="org-keyword">import</span> tensorflow <span class="org-keyword">as</span> tf
</pre>
</div>
</div>
</div>

<div id="outline-container-org6776455" class="outline-2">
<h2 id="org6776455">Simulated data</h2>
<div class="outline-text-2" id="text-org6776455">
<div class="org-src-container">
<pre class="src src-ipython" id="org6bc665f"><span class="org-keyword">with</span> nwas.simulation.simulation(p=1000, pve=0.5, annotation_params=[(10, 1)], seed=0) <span class="org-keyword">as</span> s:
  <span class="org-variable-name">x</span>, <span class="org-variable-name">y</span> = s.sample_gaussian(n=500)
  <span class="org-variable-name">x</span> = x.astype(<span class="org-string">'float32'</span>)
  <span class="org-variable-name">y</span> = y.reshape(-1, 1).astype(<span class="org-string">'float32'</span>)
</pre>
</div>
</div>
</div>

<div id="outline-container-orgae34bdc" class="outline-2">
<h2 id="orgae34bdc">Coordinate ascent</h2>
<div class="outline-text-2" id="text-orgae34bdc">
<p>
Between the old approximation and this approximation, the only difference is
\(KL\left(q(z)\Vert p(z)\right)\), but it has similar form. This suggests
that the same update would work.
</p>

<p>
The real problem is that the optimization is now constrained to have \(\pi\)
on the probability simplex, which isn't captured in the objective function.
</p>

<p>
Normalizing \(\pi_k\) after each update does not increase the ELBO.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">elbo</span>(x, y, pip, mean, var, effect_var, residual_var):
  <span class="org-variable-name">n</span>, <span class="org-variable-name">p</span> = x.shape
  <span class="org-variable-name">genetic_value_mean</span> = np.dot(x, (pip * mean).<span class="org-builtin">sum</span>(axis=0, keepdims=<span class="org-constant">True</span>).T)
  <span class="org-variable-name">genetic_value_var</span> = np.dot(np.square(x), (pip * var.T + pip * (1 - pip) * np.square(mean)).<span class="org-builtin">sum</span>(axis=0, keepdims=<span class="org-constant">True</span>).T)
  <span class="org-variable-name">llik</span> = -.5 / residual_var * (np.square(y - genetic_value_mean) - genetic_value_var).<span class="org-builtin">sum</span>()
  <span class="org-comment-delimiter"># </span><span class="org-comment">Assume prior probability 1/p for each variant</span>
  <span class="org-variable-name">kl_z</span> = (pip * (np.log(pip) + np.log(p))).<span class="org-builtin">sum</span>()
  <span class="org-variable-name">kl_b</span> = .5 * (pip * (1 + np.log(effect_var * residual_var) - np.log(var.T) + (np.square(mean) + var.T) / (effect_var * residual_var))).<span class="org-builtin">sum</span>()
  <span class="org-keyword">return</span> llik, kl_z, kl_b

<span class="org-keyword">def</span> <span class="org-function-name">coordinate_ascent</span>(x, y, effect_var, residual_var, l=5, num_epochs=200):
  <span class="org-variable-name">n</span>, <span class="org-variable-name">p</span> = x.shape
  <span class="org-variable-name">pi</span> = np.ones((p, 1))
  <span class="org-variable-name">d</span> = np.einsum(<span class="org-string">'ij,ij-&gt;j'</span>, x, x).reshape(-1, 1)
  <span class="org-variable-name">xy</span> = x.T.dot(y)
  <span class="org-variable-name">pip</span> = np.zeros((l, p))
  <span class="org-variable-name">mean</span> = np.zeros((l, p))
  <span class="org-comment-delimiter"># </span><span class="org-comment">Make sure everything is two dimensional to catch numpy broadcasting gotchas</span>
  <span class="org-variable-name">var</span> = (effect_var * residual_var / (effect_var * d + 1)).reshape(-1, 1)
  <span class="org-variable-name">eta</span> = np.dot(x, (pip * mean).<span class="org-builtin">sum</span>(axis=0, keepdims=<span class="org-constant">True</span>).T)
  <span class="org-variable-name">elbo_</span> = <span class="org-constant">None</span>
  <span class="org-variable-name">trace</span> = []
  <span class="org-keyword">for</span> epoch <span class="org-keyword">in</span> <span class="org-builtin">range</span>(num_epochs):
    <span class="org-keyword">for</span> k <span class="org-keyword">in</span> <span class="org-builtin">range</span>(l):
      <span class="org-variable-name">eta</span> -= np.dot(x, (pip * mean)[k:k + 1].T)
      <span class="org-variable-name">mean</span>[k:k + 1] = (var / residual_var * (xy - x.T.dot(eta))).T
      <span class="org-variable-name">pip</span>[k:k + 1] = (pi * np.exp(.5 * (np.log(var / (effect_var * residual_var)) + np.square(mean[k:k + 1].T) / var))).T
      <span class="org-variable-name">pip</span>[k] /= pip[k].<span class="org-builtin">sum</span>()
      <span class="org-variable-name">eta</span> += np.dot(x, (pip * mean)[k:k + 1].T)
    <span class="org-variable-name">llik</span>, <span class="org-variable-name">kl_z</span>, <span class="org-variable-name">kl_b</span> = elbo(x, y, pip, mean, var, effect_var, residual_var)
    <span class="org-variable-name">update</span> = llik - kl_z - kl_b
    trace.append([update, llik, kl_z, kl_b])
    <span class="org-variable-name">elbo_</span> = update
  <span class="org-keyword">return</span> {<span class="org-string">'pip'</span>: pip,
          <span class="org-string">'mean'</span>: pip * mean,
          <span class="org-string">'var'</span>: var,
          <span class="org-string">'elbo'</span>: elbo_,
          <span class="org-string">'trace'</span>: trace}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython" id="orgb17071b"><span class="org-variable-name">opt</span> = coordinate_ascent(x, y, effect_var=1, residual_var=s.residual_var, l=5, num_epochs=10)
pd.DataFrame(opt[<span class="org-string">'trace'</span>], columns=[<span class="org-string">'elbo'</span>, <span class="org-string">'llik'</span>, <span class="org-string">'kl_z'</span>, <span class="org-string">'kl_b'</span>])
</pre>
</div>

<pre class="example">
       elbo        llik       kl_z       kl_b
0 -268.129734 -226.243188  25.352915  16.533631
1 -268.576104 -226.869986  25.130829  16.575289
2 -268.640320 -226.956885  25.105590  16.577845
3 -268.635618 -226.929528  25.125407  16.580682
4 -268.617950 -226.879634  25.153811  16.584505
5 -268.593716 -226.816238  25.188131  16.589346
6 -268.535093 -226.682496  25.255828  16.596769
7 -268.360134 -226.292446  25.456290  16.611398
8 -268.081829 -225.462570  25.979514  16.639745
9 -268.019034 -224.800028  26.550398  16.668607
</pre>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">plot_categorical_slab_fit</span>(s, opt):
  plt.clf()
  <span class="org-variable-name">q</span> = np.logical_or(s.theta != 0, opt[<span class="org-string">'pip'</span>].<span class="org-builtin">sum</span>(axis=0) &gt; 0.1)
  <span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(3, 1, gridspec_kw={<span class="org-string">'height_ratios'</span>: [1, 1, opt[<span class="org-string">'pip'</span>].shape[0]]})
  fig.set_size_inches(6, 8)
  ax[0].bar(np.arange(np.<span class="org-builtin">sum</span>(q)), s.theta[q])
  ax[0].set_xticks(np.arange(q.<span class="org-builtin">sum</span>()))
  ax[0].set_ylabel(<span class="org-string">'True effect size'</span>)
  ax[1].bar(np.arange(np.<span class="org-builtin">sum</span>(q)), opt[<span class="org-string">'mean'</span>].<span class="org-builtin">sum</span>(axis=0)[q])
  ax[1].set_ylabel(<span class="org-string">'Estimated effect size'</span>)
  <span class="org-variable-name">im</span> = ax[2].imshow(opt[<span class="org-string">'pip'</span>][:,q], cmap=<span class="org-string">'Reds'</span>, vmin=0, vmax=1)
  ax[2].set_yticks(np.arange(opt[<span class="org-string">'pip'</span>].shape[0]))
  ax[2].set_ylabel(<span class="org-string">'Posterior inclusion probability'</span>)
  ax[2].set_xlabel(<span class="org-string">'True and false positive variants'</span>)
  plt.colorbar(im, ax=ax[2], orientation=<span class="org-string">'horizontal'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plot_categorical_slab_fit(s, opt)
</pre>
</div>


<div class="figure">
<p><img src="coordinate-ascent-opt.png" alt="coordinate-ascent-opt.png" />
</p>
</div>

<p>
<b>TODO:</b> why is the log likelihood decreasing in the second round of updates?
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">opt</span> = coordinate_ascent(x, y, l=5, effect_var=1, residual_var=s.residual_var, num_epochs=1)
plot_categorical_slab_fit(s, opt)
</pre>
</div>


<div class="figure">
<p><img src="coordinate-ascent-1-iter.png" alt="coordinate-ascent-1-iter.png" />
</p>
</div>
</div>
</div>

<div id="outline-container-orge47c8d0" class="outline-2">
<h2 id="orge47c8d0">Harder simulation</h2>
<div class="outline-text-2" id="text-orge47c8d0">
<p>
Use real LD from <a href="http://www.geuvadis.org/web/geuvadis/rnaseq-project">GEUVADIS genotypes</a>.
</p>

<div class="org-src-container">
<pre class="src src-shell" id="orgc3346dd">srun --partition=broadwl --mem=8G plink --memory 8000 --vcf /project/compbio/geuvadis/genotypes/GEUVADIS.chr1.PH1PH2_465.IMPFRQFILT_BIALLELIC_PH.annotv2.genotypes.vcf.gz --make-bed --out 1
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython" id="orgbee1434"><span class="org-variable-name">geuvadis_chr1</span> = pyplink.PyPlink(os.path.join(os.getenv(<span class="org-string">'SCRATCH'</span>), <span class="org-string">'spikeslab'</span>, <span class="org-string">'1'</span>))
<span class="org-variable-name">x</span> = np.zeros((geuvadis_chr1.get_nb_samples(), s.p), dtype=np.float32)
<span class="org-keyword">for</span> i, (_, geno) <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(geuvadis_chr1):
  <span class="org-keyword">if</span> i &gt;= x.shape[1]:
    <span class="org-keyword">break</span>
  <span class="org-variable-name">x</span>[:,i] = geno.astype(np.float32)
<span class="org-variable-name">x</span> = np.ma.masked_equal(x, -1)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython" id="org5daca47"><span class="org-keyword">def</span> <span class="org-function-name">plot_ld</span>(x):
  <span class="org-variable-name">w</span> = x - x.mean(axis=0)
  <span class="org-variable-name">w</span> /= 1e-8 + w.std(axis=0)
  <span class="org-variable-name">corr</span> = w.T.dot(w) / w.shape[0]
  plt.clf()
  plt.gcf().set_size_inches(8, 8)
  plt.imshow(np.triu(corr), cmap=<span class="org-string">'RdBu_r'</span>, vmin=-1, vmax=1)
  <span class="org-variable-name">cb</span> = plt.colorbar()
  cb.ax.set_xlabel(<span class="org-string">'Correlation'</span>)

plot_ld(x)
</pre>
</div>


<div class="figure">
<p><img src="ld.png" alt="ld.png" />
</p>
</div>

<p>
Simulate a problem where causal variants almost certain to be in tight LD.
</p>

<div class="org-src-container">
<pre class="src src-ipython">s.estimate_mafs(x)
<span class="org-variable-name">annotation</span> = np.zeros(s.p)
<span class="org-variable-name">annotation</span>[600:610] = np.ones(10)
s.load_annotations(annotation)
s.sample_effects(pve=0.5, annotation_params=[(0, 1), (2, 1)], permute=<span class="org-constant">True</span>)
<span class="org-variable-name">x</span> = (x - x.mean(axis=0)).filled(0)
<span class="org-variable-name">y</span> = s.compute_liabilities(x).reshape(-1, 1)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython" id="org14c30b5"><span class="org-variable-name">opt</span> = coordinate_ascent(x, y, effect_var=1, residual_var=s.residual_var, l=5, num_epochs=10)
pd.DataFrame(opt[<span class="org-string">'trace'</span>], columns=[<span class="org-string">'elbo'</span>, <span class="org-string">'llik'</span>, <span class="org-string">'kl_z'</span>, <span class="org-string">'kl_b'</span>])
</pre>
</div>

<pre class="example">
       elbo        llik      kl_z       kl_b
0 -219.908757 -200.251698  8.487061  11.169998
1 -219.875655 -200.537041  8.216742  11.121872
2 -220.072682 -200.806115  8.163078  11.103489
3 -220.145278 -200.903071  8.147945  11.094261
4 -220.163084 -200.926393  8.144994  11.091697
5 -220.164622 -200.928121  8.145114  11.091388
6 -220.163479 -200.926487  8.145488  11.091504
7 -220.162768 -200.925524  8.145652  11.091592
8 -220.162557 -200.925248  8.145688  11.091621
9 -220.162536 -200.925225  8.145687  11.091625
</pre>

<div class="org-src-container">
<pre class="src src-ipython">plot_categorical_slab_fit(s, opt)
</pre>
</div>


<div class="figure">
<p><img src="categorical-slab-geuvadis.png" alt="categorical-slab-geuvadis.png" />
</p>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2017-11-02 Thu 15:39</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
