<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<head>
<!-- 2017-10-04 Wed 22:03 -->
<meta http-equiv="Content-Type" content="text/html;charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<title>Two-stage regression</title>
<meta name="generator" content="Org mode" />
<meta name="author" content="Abhishek Sarkar" />
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Two-stage regression</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#orga39774c">1. Introduction</a></li>
<li><a href="#org1331c2f">2. Simulation setup</a></li>
<li><a href="#orgfc0bf5c">3. Joint model</a></li>
<li><a href="#org5fd20e0">4. Variational approximation</a></li>
<li><a href="#orge2f7e14">5. Model fitting</a></li>
<li><a href="#org160e68d">6. Model evaluation</a></li>
<li><a href="#org795971c">7. Limitations</a></li>
</ul>
</div>
</div>

<div id="outline-container-orga39774c" class="outline-2">
<h2 id="orga39774c"><span class="section-number-2">1</span> Introduction</h2>
<div class="outline-text-2" id="text-1">
<p>
Suppose we generated data from both mediated and unmediated effects:
</p>

<p>
\[ y^1 = x^1 w v + x^1 u + e \]
</p>

<p>
If we require that \(w\) explains the gene expression observations, then can
we accurately estimate \(v\) even if \(u\) is correlated with \(v\)?
</p>

<p>
It is biologically plausible for a variant to have both a mediated and an
unmediated effect?
</p>

<p>
Do we have any documented examples of this?
</p>

<p>
Can we estimate how much variance is explained by unmediated effects?
</p>

<p>
Can we eliminate this possibility by considering a large enough region of the
genome (enough genes)? Can we eliminate it by regressing out the rest of the
genome?
</p>
</div>
</div>

<div id="outline-container-org1331c2f" class="outline-2">
<h2 id="org1331c2f"><span class="section-number-2">2</span> Simulation setup</h2>
<div class="outline-text-2" id="text-2">
<ul class="org-ul">
<li>Generate random genotypes in linkage equilibrium</li>

<li>Generate Gaussian gene expression from a linear model (PVE = 0.5)</li>
</ul>

<div class="LaTeX">
<p>
\[ w_j \mid \text{causal} ~ N(0, 1) \]
\[ g^0_{ik} \mid \text{causal} ~ x^0 w + e \]
</p>

</div>

<ul class="org-ul">
<li>Generate non-causal gene expression by sampling from a Gaussian with scale
equal to the simulated expression phenotypic variance.</li>
</ul>

<p>
\[ g_{ik} \mid \text{not causal} ~ N(0, V[g \mid \causal]) \]
</p>

<ul class="org-ul">
<li>Generate GWAS expression from the same linear model</li>
</ul>

<p>
\[ g^1_{ik} \mid \text{causal} ~ x^1 w + e \]
</p>

<ul class="org-ul">
<li>Generate pleiotropic unmediated effect (same variant, independent
effect on phenotype). Scale to get the desired PVE.</li>
</ul>

<p>
\[ u \mid \text{causal} = N(0, 1) \]
</p>

<ul class="org-ul">
<li>Generate Gaussian phenotypes using the unmediated effect and adding noise
to get the desired PVE.</li>
</ul>

<p>
\[ y^1 \mid N(x^1 u, \sigma^2) \]
</p>
</div>
</div>

<div id="outline-container-orgfc0bf5c" class="outline-2">
<h2 id="orgfc0bf5c"><span class="section-number-2">3</span> Joint model</h2>
<div class="outline-text-2" id="text-3">
<div class="org-src-container">
<pre><code class="src src-python">ed.set_seed(0)

<span style="color: #00ffff; font-style: italic;"># </span><span style="color: #00ffff; font-style: italic;">Data</span>
<span style="color: #0000ee;">x0</span> = tf.placeholder(tf.float32)
<span style="color: #0000ee;">x1</span> = tf.placeholder(tf.float32)

<span style="color: #00ffff; font-style: italic;"># </span><span style="color: #00ffff; font-style: italic;">eQTL effects</span>
<span style="color: #0000ee;">logodds_w</span> = Normal(loc=tf.constant(-10.0), scale=tf.ones(1))
<span style="color: #0000ee;">scale_w</span> = Normal(loc=tf.zeros(1), scale=tf.ones(1))
<span style="color: #0000ee;">w</span> = SpikeSlab(logodds=logodds_w, loc=tf.zeros([p, m]), scale=scale_w)
<span style="color: #00ffff; font-style: italic;"># </span><span style="color: #00ffff; font-style: italic;">This is a dummy which gets swapped out in the inference</span>
<span style="color: #0000ee;">eta0</span> = LocalReparameterization(Normal(tf.matmul(x0, w), 1.0))
<span style="color: #0000ee;">g0</span> = NormalWithSoftplusScale(
    loc=eta0, scale=tf.Variable(tf.random_normal([1])))

<span style="color: #00ffff; font-style: italic;"># </span><span style="color: #00ffff; font-style: italic;">Mediated gene effects</span>
<span style="color: #0000ee;">logodds_v</span> = Normal(loc=tf.constant(-10.0), scale=tf.ones(1))
<span style="color: #0000ee;">scale_v</span> = Normal(loc=tf.zeros(1), scale=tf.ones(1))
<span style="color: #0000ee;">v</span> = SpikeSlab(logodds=logodds_v, loc=tf.zeros([m, 1]), scale=scale_v)
<span style="color: #0000ee;">eta1m</span> = LocalReparameterization(Normal(tf.matmul(tf.matmul(x1, w), v), 1.0))

<span style="color: #00ffff; font-style: italic;"># </span><span style="color: #00ffff; font-style: italic;">Unmediated effects</span>
<span style="color: #0000ee;">logodds_u</span> = Normal(loc=tf.constant(-10.0), scale=tf.ones(1))
<span style="color: #0000ee;">scale_u</span> = Normal(loc=tf.zeros(1), scale=tf.ones(1))
<span style="color: #0000ee;">u</span> = SpikeSlab(logodds=logodds_u, loc=tf.zeros([p, 1]), scale=scale_u)
<span style="color: #0000ee;">eta1u</span> = LocalReparameterization(Normal(tf.matmul(x1, u), 1.0))

<span style="color: #0000ee;">y1</span> = NormalWithSoftplusScale(loc=eta1m + eta1u, scale=tf.Variable(0.0))
</code></pre>
</div>
</div>
</div>

<div id="outline-container-org5fd20e0" class="outline-2">
<h2 id="org5fd20e0"><span class="section-number-2">4</span> Variational approximation</h2>
<div class="outline-text-2" id="text-4">
<div class="org-src-container">
<pre><code class="src src-python"><span style="color: #0000ee;">q_logodds_w</span> = Normal(loc=tf.Variable(tf.random_normal([1])),
                     scale=tf.Variable(tf.random_normal([1])))
<span style="color: #0000ee;">q_logodds_v</span> = Normal(loc=tf.Variable(tf.random_normal([1])),
                     scale=tf.Variable(tf.random_normal([1])))
<span style="color: #0000ee;">q_logodds_u</span> = Normal(loc=tf.Variable(tf.random_normal([1])),
                     scale=tf.Variable(tf.random_normal([1])))
<span style="color: #0000ee;">q_scale_w</span> = Normal(loc=tf.Variable(tf.random_normal([1])),
                   scale=tf.Variable(tf.random_normal([1])))
<span style="color: #0000ee;">q_scale_v</span> = Normal(loc=tf.Variable(tf.random_normal([1])),
                   scale=tf.Variable(tf.random_normal([1])))
<span style="color: #0000ee;">q_scale_u</span> = Normal(loc=tf.Variable(tf.random_normal([1])),
                   scale=tf.Variable(tf.random_normal([1])))

<span style="color: #00cd00;">import</span> scipy.linalg
initial_w, *<span style="color: #0000ee;">_</span> = scipy.linalg.lstsq(x_ref, g_ref)

<span style="color: #0000ee;">q_w</span> = SpikeSlab(logodds=tf.Variable(tf.zeros([p, m])),
                loc=tf.Variable(initial_w.astype(<span style="color: #00cdcd;">'float32'</span>)),
                scale=tf.Variable(tf.zeros([p, m])))
<span style="color: #0000ee;">q_eta0</span> = LocalReparameterization(
    Normal(loc=tf.matmul(x0, q_w.mean()),
           scale=tf.sqrt(tf.matmul(tf.square(x0), q_w.variance()))))
</code></pre>
</div>

<p>
We need to do some work to get the reparameterized distribution \(q(X w
  v)\). As previously derived (Brown 1977), if b, c are \(n\)-dimensional
Gaussian then:
</p>

<p>
\[ E[b' c] = E[b]' E[c] \]
</p>

<p>
\[ V[b' c] = E[b]' Cov(c, c) E[b] + E[c]' Cov(b, b) + E[c]  + \Tr(Cov(b, b) Cov(c, c)) \]
</p>

<p>
Here, we need moments of a stochastic matrix-vector product. However, under the
variational approximation, all of the elements are independent, simplifying the
derivation. Let \(\eta = X w\). Then considering each row \(\eta_i\), we can
simply apply the above result to get:
</p>

<p>
\[ E_q[\eta_i v] = E_q[\eta_i] E_q[v] \]
</p>

<p>
\[ V_q[\eta_i v] = E_q[\eta_i] \diag(V_q[v]) E_q[\eta_i]' + E_q[v]' \diag(V_q[\eta_i]) E_q[v] + V_q[\eta_i]' V_q[v] \]
</p>

<div class="org-src-container">
<pre><code class="src src-python"><span style="color: #0000ee;">q_v</span> = SpikeSlab(logodds=tf.Variable(tf.zeros([m, 1])),
                loc=tf.Variable(tf.random_normal([m, 1])),
                scale=tf.Variable(tf.zeros([m, 1])))
<span style="color: #00ffff; font-style: italic;"># </span><span style="color: #00ffff; font-style: italic;">Conviently keep the necessary mean and variance around</span>
<span style="color: #0000ee;">q_eta1</span> = Normal(loc=tf.matmul(x1, q_w.mean()),
                scale=tf.sqrt(tf.matmul(tf.square(x1), q_w.variance())))
<span style="color: #0000ee;">var</span> = (tf.reduce_sum(tf.square(q_eta1.mean()) *
                     tf.transpose(q_v.variance()), axis=1, keep_dims=<span style="color: #00cdcd;">True</span>) +
       tf.reduce_sum(tf.transpose(tf.square(q_v.mean())) *
                     q_eta1.variance(), axis=1, keep_dims=<span style="color: #00cdcd;">True</span>) +
       tf.matmul(q_eta1.variance(), q_v.variance()))
<span style="color: #0000ee;">q_eta1m</span> = LocalReparameterization(
    Normal(loc=tf.matmul(tf.matmul(x1, q_w.mean()), q_v.mean()),
           scale=tf.sqrt(var)))

<span style="color: #0000ee;">q_u</span> = SpikeSlab(logodds=tf.Variable(tf.zeros([p, 1])),
                loc=tf.Variable(tf.zeros([p, 1])),
                scale=tf.Variable(tf.zeros([p, 1])))
<span style="color: #0000ee;">q_eta1u</span> = LocalReparameterization(
    Normal(loc=tf.matmul(x1, q_u),
           scale=tf.sqrt(tf.matmul(tf.square(x1), q_u.variance()))))
</code></pre>
</div>
</div>
</div>

<div id="outline-container-orge2f7e14" class="outline-2">
<h2 id="orge2f7e14"><span class="section-number-2">5</span> Model fitting</h2>
<div class="outline-text-2" id="text-5">
<div class="org-src-container">
<pre><code class="src src-python"><span style="color: #0000ee;">inference</span> = ed.ReparameterizationKLKLqp(
    latent_vars={
        logodds_w: q_logodds_w,
        logodds_v: q_logodds_v,
        logodds_u: q_logodds_u,
        scale_w: q_scale_w,
        scale_v: q_scale_v,
        scale_u: q_scale_u,
        w: q_w,
        v: q_v,
        eta0: q_eta0,
        eta1m: q_eta1m,
        eta1u: q_eta1u,
    },
    data={
        x0: x_ref,
        g0: g_ref,
        x1: x_gwas,
        y1: y_gwas,
    })
inference.run(n_iter=2000, optimizer=<span style="color: #00cdcd;">'rmsprop'</span>)
</code></pre>
</div>
</div>
</div>
<div id="outline-container-org160e68d" class="outline-2">
<h2 id="org160e68d"><span class="section-number-2">6</span> Model evaluation</h2>
<div class="outline-text-2" id="text-6">
<div class="org-src-container">
<pre><code class="src src-python"><span style="color: #00cd00;">import</span> matplotlib.gridspec
<span style="color: #00cd00;">import</span> matplotlib.pyplot <span style="color: #00cd00;">as</span> plt

<span style="color: #0000ee;">sess</span> = ed.get_session()
<span style="color: #0000ee;">est_w</span> = sess.run(q_w.pip)
<span style="color: #0000ee;">est_v</span> = sess.run(q_v.pip)
<span style="color: #0000ee;">est_u</span> = sess.run(q_u.pip)

plt.switch_backend(<span style="color: #00cdcd;">'pdf'</span>)
<span style="color: #0000ee;">gs</span> = matplotlib.gridspec.GridSpec(2, 2, width_ratios=[100, 1])
<span style="color: #0000ee;">norm</span> = matplotlib.colors.NoNorm(0, 1)
<span style="color: #0000ee;">fig</span> = plt.gcf()
fig.set_size_inches(8, 2)
plt.clf()
plt.subplot(gs[0])
plt.imshow(est_w.T, cmap=<span style="color: #00cdcd;">'Greys'</span>, norm=norm)
<span style="color: #0000ee;">ax</span> = plt.gca()
ax.set_xticks([])
ax.set_yticks([])
ax.set_xticklabels([])
ax.set_yticklabels([])
ax.set_xlabel(<span style="color: #00cdcd;">''</span>)
ax.set_ylabel(<span style="color: #00cdcd;">'Genes'</span>)

plt.subplot(gs[1])
plt.imshow(est_v, cmap=<span style="color: #00cdcd;">'Greys'</span>, norm=norm)
<span style="color: #0000ee;">ax</span> = plt.gca()
ax.set_xticks([])
ax.set_yticks([])
ax.set_xticklabels([])
ax.set_yticklabels([])

plt.subplot(gs[2])
plt.imshow(est_u.T, cmap=<span style="color: #00cdcd;">'Greys'</span>, norm=norm)
<span style="color: #0000ee;">ax</span> = plt.gca()
ax.set_xticks([])
ax.set_yticks([])
ax.set_xticklabels([])
ax.set_yticklabels([])
ax.set_xlabel(<span style="color: #00cdcd;">'Variants'</span>)

plt.savefig(<span style="color: #00cdcd;">'coefficients-pleiotropy'</span>)
plt.close()
</code></pre>
</div>

<p>
<a href="coefficients-pleiotropy.pdf">coefficients-pleiotropy.pdf</a>
</p>
</div>
</div>
<div id="outline-container-org795971c" class="outline-2">
<h2 id="org795971c"><span class="section-number-2">7</span> Limitations</h2>
<div class="outline-text-2" id="text-7">
<p>
\(u\) correlated with \(v\) is the well studied pleiotropy problem
in Mendelian randomization.
</p>

<p>
Unfortunately, prior work (by others, and us!) on this problem suggests it
can't be solved in one model. Starting from the Mendelian randomization
perspective: if the goal is to estimate the causal effect \(v\), then we have
to first estimate \(w\), then estimate \(v\) fixing \(X w\).
</p>

<p>
Now, in the case where \(u\) is correlated with \(v\), this method is biased
and requires correction. The relevant method is <a href="https://www.ncbi.nlm.nih.gov/pubmed/26050253">MR-Egger</a>, which has also been
extended to <a href="https://arxiv.org/abs/1708.00272">multivariate case</a>. The key idea is to fit weighted least squares,
where the weights come from the standard error of the direct associations.
</p>

<p>
\[ (w v + u) = w \tilde{v} + b \]
</p>

<p>
Open questions:
</p>

<ol class="org-ol">
<li>If instruments are correlated (thinking about strong LD within a locus),
what breaks in MR-Egger?</li>
<li>If we replace OLS with a posterior mean assuming the spike-and-slab
prior, what happens?</li>
<li>The key assumption (INSIDE) is that \(Cov(w v, u) = 0\). Is this plausible
for the case of <i>cis</i>-regulatory variants within a single locus?</li>
<li>The multivariate extension still requires an independence assumption
between the different mediators. If we fit it using SSB, what happens?</li>
</ol>

<p>
To make causal claims (about mediation), we further need to remove
<i>trans</i>-effects and reverse causal effects on gene expression.
</p>

<p>
We can do the first using half-sibling regression: regress observed genes
expression against control gene expression, where control genes are on other
chromosomes.
</p>

<p>
We can do the second using a random effects approach. Suppose we regress gene
expression against both genotype and phenotype, assuming a linear mixed
model. Treat genotype effects as random by building a kernel matrix the rest
of the genome, and treat expression effects as fixed (estimated).
</p>

<p>
This could be done using BSLMM, or variational BSLMM as hinted by Peter.
</p>

<p>
This still assumes gene expression is Gaussian. It should be trivial to build
a negative binomial observation model over a latent linear model using
Edward.
</p>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2017-10-04 Wed 22:03</p>
<p class="creator">Emacs 25.1.1 (Org mode 8.2.10)</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
